File: ios/LynxShellKit/Runtime/ShellGlobalPropsFactory.swift
import UIKit

/** 构造跨端保留字段；系统字段最后写入，避免被页面参数伪造。 */
enum ShellGlobalPropsFactory {
    static func make(
        for view: UIView,
        request: LynxPageRequest,
        pageId: String? = nil,
        sessionId: String? = nil,
        bundleMetadata: [String: Any]? = nil,
        localeState: LynxLocaleState? = nil,
        layoutSnapshot: ShellLayoutSnapshot? = nil
    ) -> [String: Any] {
        var props = request.globalProps
        let info = Bundle.main.infoDictionary ?? [:]
        let resolvedLocaleState = localeState ?? ShellLocaleStore.current()
        let resolvedLayoutSnapshot = layoutSnapshot ?? ShellLayoutSnapshot(
            revision: 0,
            timestampMillis: Int64(Date().timeIntervalSince1970 * 1000),
            measurement: ShellLayoutSnapshot.measure(for: view)
        )

        props["platform"] = "ios"
        // Sparkling Playground 的旧字段名。保留当前壳字段，同时提供兼容别名。
        props["os"] = "ios"
        props["theme"] = view.traitCollection.userInterfaceStyle == .dark ? "Dark" : "Light"
        props["frontendTheme"] = "system"
        props["systemVersion"] = UIDevice.current.systemVersion
        props["appVersion"] = info["CFBundleShortVersionString"] as? String ?? ""
        props["buildNumber"] = info["CFBundleVersion"] as? String ?? ""
        applyLayout(resolvedLayoutSnapshot, to: &props)
        applyLocale(resolvedLocaleState, to: &props)
        // 页面应读取原生容器最终采用的 chrome 状态，避免原始 query 与 options 合并后
        // 留下过期值。其他业务参数保持不变，宿主保留字段在这里统一覆盖。
        var queryItems: [String: Any] = [:]
        if let values = props["queryItems"] as? [String: Any] {
            queryItems = values
        } else if let values = props["queryItems"] as? [String: String] {
            // Native Tab 描述通常使用 [String: String]；显式桥接后再合并宿主字段，避免
            // Swift 泛型字典不协变导致 native_tab_id 被静默丢弃。
            queryItems = values.mapValues { $0 as Any }
        } else if let values = props["queryItems"] as? NSDictionary {
            values.forEach { key, value in
                if let key = key as? String {
                    queryItems[key] = value
                }
            }
        }
        queryItems["fullscreen"] = request.fullscreen ? "1" : "0"
        queryItems["hide_nav_bar"] = request.showNavigationBar ? "0" : "1"
        queryItems["hide_status_bar"] = request.hideStatusBar ? "1" : "0"
        queryItems["trans_status_bar"] =
            request.fullscreen && !request.hideStatusBar ? "1" : "0"
        queryItems["locale"] = resolvedLocaleState.locale
        queryItems["language"] = resolvedLocaleState.language
        queryItems["localeRevision"] = resolvedLocaleState.revision
        props["queryItems"] = queryItems
        // v1 跨端契约：containerID 兼容旧 Shell 字段，但身份必须按“页面实例”唯一，
        // 不能再使用 Bundle hash，否则同一个 Bundle 多次 push 会互相覆盖消息目标。
        let resolvedPageId = pageId?.isEmpty == false ? pageId! : request.resolvedRouteKey
        let resolvedSessionId = sessionId?.isEmpty == false ? sessionId! : ""
        props["containerID"] = resolvedPageId
        props["__lynxRouterContainerId"] = resolvedPageId
        props["__lynxRouterPageId"] = resolvedPageId
        props["__lynxRouterPageKey"] = request.resolvedRouteKey
        props["__lynxRouterSessionId"] = resolvedSessionId
        props["__lynxRouterNavigationModel"] = "native_page_stack"
        props["__lynxRouterPlatformContainer"] = "uikit_view_controller"
        props["__lynxRouterParams"] = queryItems
        if let bundleMetadata {
            props["__lynxBundleMeta"] = bundleMetadata
        } else {
            props.removeValue(forKey: "__lynxBundleMeta")
        }
        if let nativeTransition = request.nativeTransition {
            // 必须在 loadTemplate 之前注入，目标页首屏即可读取 transactionID 并按需
            // 调用 NativeModules.LynxShellModule.markTransitionReady。
            props["nativeTransition"] = nativeTransition.dictionary
        } else {
            props.removeValue(forKey: "nativeTransition")
        }
        return props
    }

    static func applyLocale(_ state: LynxLocaleState, to props: inout [String: Any]) {
        props["__lynxShellLocale"] = state.dictionary
        props["locale"] = state.locale
        props["effectiveLocale"] = state.locale
        props["formatLocale"] = state.locale
        props["language"] = state.language
        props["appLocale"] = state.appLocale ?? NSNull()
        props["appLocaleOverride"] = state.appLocale ?? NSNull()
        props["appLanguage"] = state.language
        props["systemLocale"] = state.systemLocale
        props["localeRevision"] = state.revision
        props["localeSource"] = state.source
        props["localeStatus"] = state.status
        props["direction"] = state.direction
    }

    static func applyLayout(_ snapshot: ShellLayoutSnapshot, to props: inout [String: Any]) {
        let measurement = snapshot.measurement
        props["__lynxShellLayout"] = snapshot.dictionary
        props["screenWidth"] = measurement.screenWidth
        props["screenHeight"] = measurement.screenHeight
        props["viewportWidth"] = measurement.viewportWidth
        props["viewportHeight"] = measurement.viewportHeight
        props["density"] = measurement.density
        props["safeAreaTop"] = measurement.safeAreaTop
        props["safeAreaBottom"] = measurement.safeAreaBottom
        props["safeAreaLeft"] = measurement.safeAreaLeft
        props["safeAreaRight"] = measurement.safeAreaRight
        props["topHeight"] = measurement.safeAreaTop
        props["bottomHeight"] = measurement.safeAreaBottom
        props["statusBarHeight"] = measurement.safeAreaTop
        props["navigationBarHeight"] = measurement.safeAreaBottom
        props["isNotchScreen"] = measurement.safeAreaTop > 20
        props["layoutRevision"] = snapshot.revision
    }
}

File: android/lynx-shell/src/main/java/com/example/lynxshell/runtime/ShellGlobalPropsFactory.kt
package com.example.lynxshell.runtime

import android.app.Activity
import android.content.res.Configuration
import android.os.Build
import android.view.View
import com.example.lynxshell.model.LynxPageRequest
import com.example.lynxshell.routing.LynxNavigator
import com.example.lynxshell.transition.LynxTransitionIntent
import com.example.lynxshell.util.JsonObjectCodec
import com.lynx.tasm.LynxColorScheme

/** 构造两端约定的宿主全局参数；系统保留字段不允许页面覆盖。 */
object ShellGlobalPropsFactory {
    /** 将当前 Activity 的有效夜间模式映射为 Lynx 4.0 两态颜色枚举。 */
    fun resolveColorScheme(activity: Activity): LynxColorScheme =
        if (isDarkTheme(activity)) LynxColorScheme.DARK else LynxColorScheme.LIGHT

    /** 与页面 GlobalProps 使用同一份主题判定，避免引擎和页面出现不同步。 */
    fun resolveThemeName(activity: Activity): String =
        if (isDarkTheme(activity)) "Dark" else "Light"

    private fun isDarkTheme(activity: Activity): Boolean =
        activity.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK ==
            Configuration.UI_MODE_NIGHT_YES

    /** 供 Builder 和布局协调器读取同一份 WindowMetrics。 */
    fun captureLayout(activity: Activity, view: View? = null): LynxLayoutSnapshot =
        LynxLayoutSnapshot.capture(activity, view)

    /** 构造不依赖具体路由身份的环境字段；GlobalProps 更新会合并这些保留字段。 */
    fun createEnvironment(
        activity: Activity,
        view: View? = null,
        snapshot: LynxLayoutSnapshot = captureLayout(activity, view),
        layoutRevision: Long = 0L,
        locale: LynxLocaleState = LynxLocaleStore.current(activity),
    ): HashMap<String, Any> {
        val props = snapshot.toGlobalProps(layoutRevision)
        props["platform"] = "android"
        props["os"] = "android"
        props["theme"] = resolveThemeName(activity)
        props["frontendTheme"] = "system"
        props["systemVersion"] = Build.VERSION.RELEASE
        props["locale"] = locale.effectiveLocale
        props["language"] = locale.language
        props["appLanguage"] = locale.language
        props["appLocale"] = locale.appLocale ?: org.json.JSONObject.NULL
        props["appLocaleOverride"] = locale.appLocale ?: org.json.JSONObject.NULL
        props["systemLocale"] = locale.systemLocale
        props["localeSource"] = locale.source
        props["localeStatus"] = locale.status
        props["localeRevision"] = locale.revision
        props["direction"] = locale.direction
        props["formatLocale"] = locale.effectiveLocale
        props["__lynxShellLocale"] = locale.toGlobalMap()
        props["layoutCapabilities"] = hashMapOf<String, Any>(
            "windowMetrics" to "supported",
            "viewportMetrics" to if (snapshot.viewportWidthPx != null && snapshot.viewportHeightPx != null) {
                "supported"
            } else {
                "pending"
            },
            "foldStatus" to snapshot.foldingCapability,
            "creaseGeometry" to if (snapshot.foldingFeature == null) "unavailable" else "supported",
        )
        val packageInfo = activity.packageManager.getPackageInfo(activity.packageName, 0)
        props["appVersion"] = packageInfo.versionName ?: ""
        props["buildNumber"] = if (Build.VERSION.SDK_INT >= 28) {
            packageInfo.longVersionCode.toString()
        } else {
            @Suppress("DEPRECATION") packageInfo.versionCode.toString()
        }
        return props
    }

    fun create(
        activity: Activity,
        request: LynxPageRequest,
        bundleMetadata: Map<String, Any>? = null,
        initialLayout: LynxLayoutSnapshot? = null,
    ): HashMap<String, Any> {
        val props = JsonObjectCodec.toMap(request.globalPropsJson, "globalProps")
        props.putAll(
            createEnvironment(
                activity = activity,
                snapshot = initialLayout ?: captureLayout(activity),
            ),
        )
        // 页面应读取“原生最终采用”的 chrome 状态，而不是调用方可能遗漏或互相冲突的
        // 原始参数。保留其他 queryItems，并覆盖这四个宿主保留字段。
        val queryItems = hashMapOf<String, Any>()
        (props["queryItems"] as? Map<*, *>)?.forEach { (key, value) ->
            if (key is String && value != null) {
                queryItems[key] = value
            }
        }
        queryItems["fullscreen"] = if (request.fullscreen) "1" else "0"
        queryItems["hide_nav_bar"] = if (request.showToolbar) "0" else "1"
        queryItems["hide_status_bar"] = if (request.hideStatusBar) "1" else "0"
        queryItems["trans_status_bar"] =
            if (request.fullscreen && !request.hideStatusBar) "1" else "0"
        val locale = LynxLocaleStore.current(activity)
        queryItems["locale"] = locale.effectiveLocale
        queryItems["language"] = locale.language
        props["queryItems"] = queryItems
        // Native Page Stack 身份必须按页面实例生成：同一个 Bundle 多次 push 不能共享
        // containerID，否则 sendToPage 会把消息误投到旧 Activity。
        val identity = (activity as? com.example.lynxshell.container.LynxShellActivity)
            ?.let(LynxNavigator::routerPageIdentity)
        val pageId = identity?.entryID ?: "lynx-shell-${request.bundleUrl.hashCode()}"
        props["containerID"] = pageId
        props["__lynxRouterContainerId"] = pageId
        props["__lynxRouterPageId"] = pageId
        props["__lynxRouterPageKey"] = identity?.routeKey ?: request.resolvedRouteKey()
        props["__lynxRouterSessionId"] = identity?.sessionID ?: ""
        props["__lynxRouterNavigationModel"] = "native_page_stack"
        props["__lynxRouterPlatformContainer"] = "android_activity"
        props["__lynxRouterParams"] = queryItems
        bundleMetadata?.let { props["__lynxBundleMeta"] = HashMap(it) }
        // 目标页在首屏 Bundle 执行前即可读取 transactionID，并据此调用 markTransitionReady。
        LynxTransitionIntent.globalProps(activity.intent)?.let {
            props["nativeTransition"] = it
        }
        return props
    }
}
